vm/src/stdlib/subprocess.rs

```diff
@@ -0,0 +1,182 @@
+use std::cell::RefCell;
+use std::fs::File;
+use std::time::Duration;
+
+use subprocess;
+
+use crate::function::{OptionalArg, PyFuncArgs};
+use crate::obj::objlist::PyListRef;
+use crate::obj::objsequence;
+use crate::obj::objstr::{self, PyStringRef};
+use crate::obj::objtype::PyClassRef;
+use crate::pyobject::{Either, IntoPyObject, PyObjectRef, PyRef, PyResult, PyValue};
+use crate::stdlib::io::io_open;
+use crate::stdlib::os::{raw_file_number, rust_file};
+use crate::vm::VirtualMachine;
+
+#[derive(Debug)]
+struct Popen {
+    process: RefCell<subprocess::Popen>,
+}
+
+impl PyValue for Popen {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("subprocess", "Popen")
+    }
+}
+
+type PopenRef = PyRef<Popen>;
+
+#[derive(FromArgs)]
+struct PopenArgs {
+    #[pyarg(positional_only)]
+    args: Either<PyStringRef, PyListRef>,
+    #[pyarg(positional_or_keyword, default = "None")]
+    stdin: Option<i64>,
+    #[pyarg(positional_or_keyword, default = "None")]
+    stdout: Option<i64>,
+    #[pyarg(positional_or_keyword, default = "None")]
+    stderr: Option<i64>,
+}
+
+impl IntoPyObject for subprocess::ExitStatus {
+    fn into_pyobject(self, vm: &VirtualMachine) -> PyResult {
+        let status: i32 = match self {
+            subprocess::ExitStatus::Exited(status) => status as i32,
+            subprocess::ExitStatus::Signaled(status) => -i32::from(status),
+            subprocess::ExitStatus::Other(status) => status as i32,
+            _ => return Err(vm.new_os_error("Unknown exist status".to_string())),
+        };
+        Ok(vm.new_int(status))
+    }
+}
+
+fn convert_redirection(arg: Option<i64>, vm: &VirtualMachine) -> PyResult<subprocess::Redirection> {
+    match arg {
+        Some(fd) => match fd {
+            -1 => Ok(subprocess::Redirection::Pipe),
+            -2 => panic!("TODO"),
+            -3 => panic!("TODO"),
+            fd => {
+                if fd < 0 {
+                    Err(vm.new_value_error(format!("Invalid fd: {}", fd)))
+                } else {
+                    Ok(subprocess::Redirection::File(rust_file(fd)))
+                }
+            }
+        },
+        None => Ok(subprocess::Redirection::None),
+    }
+}
+
+fn convert_to_file_io(file: &Option<File>, mode: String, vm: &VirtualMachine) -> PyResult {
+    match file {
+        Some(ref stdin) => io_open(
+            vm,
+            PyFuncArgs::new(
```

Comment:
There is a `From` implementation for `PyFuncArgs` for `Vec`, so you might be able to reduce this to:
```rust
io_open(
            vm,
            vec![
                    vm.new_int(raw_file_number(stdin.try_clone().unwrap())),
                    vm.new_str(mode),
            ],
)
```

Reply:
Did you mean to append a `.into()` on the `vec!` call?

Reply:
Yes, you are right. I was confused with the `invoke` function which accepts an `Into<PyFuncArgs>` type.